### DISCUSSION TITLE
API follows the OpenAPI specification.

### DISCUSSION THREAD

**Discussion by sunli829** [databendlabs/databend#3418] [Ideas]
If we use [`poem-openapi`](https://crates.io/crates/poem-openapi) to implement all HTTP APIs, we can automatically generate beautiful documents and client library in any language.
But if the current APIs is for internal use only, maybe there is no need to change it.

**Comment by ZhiHanZ** [comment]
LGTM, I think at least for query http handler, openapi is needed. I think @youngsofun has some ideas?

**Comment by youngsofun** [comment]
@sunli829 @ZhiHanZ  

the problem with the HTTP handler is it is in fact not so restful,  
HTTP is used as a carrier of "protocol" to stream result data by "page" (like MySQL protocol on TCP)

the endpoint itself is simple,  the complicity lies in that client needs to know the "protocol " to use it.
so open API does not help much, later we may need  special client SDKs for it as more features are added to it.

in addition, currently, we follow the presto API to export only one endpoint and hide many internal endpoints (via links), 
it is more special and may change later.
